uwblueprint/supportive-housing
# Integrate Authy 2FA

## Notion ticket link

[Ticket Name](https://github.com/uwblueprint/supportive-housing/issues/19)


## Implementation description
Integrated 2FA using Authy. Basic user flow goes as so:
- User tries to login using normal `/auth/login` route
- Instead of returning the authenticated user, the route will return the following body:
`{requires_two_fa: Boolean, auth_user: User or None}`
- If the user is a relief user, `requires_two_fa` will be set to True and no auth user will be returned, since we don't want an access token unless they're 2FA verified 
- Otherwise, the auth user is returned and the user logs in as normal
- Once the user types in the passcode from the Authy app, they will POST `/auth/twoFa?passcode={passcode}`
- Here, the backend calls the Verify API and verified whether the provided passcode in the query string is valid or not
- If it's valid, it returns the auth_user and logs the user in

Important note: I made the decision to 2FA every time the user has to login. An alternative was to request 2FA even if the user had previously logged in and still had a valid access/refresh token, but I figured that might get extremely annoying. 

Another note: I didn't write any tests for this yet (@Safewaan pls send help), but instead updated the current tests slightly so they don't break with these changes. I figured getting this PR up and reviewed ASAP was more important.


## Steps to test
I want to apologize in advance to whoever has to test this works
 
1. If you don't already have one, [create a new Twilio account](https://www.twilio.com/try-twilio) under your Blueprint email (the free trial suffices, and never ends so you won't be charged)
2. [Create a verify service](https://www.twilio.com/console/verify/services) in the console. You don't need to toggle any of the verification channels, as we're just using TOTP
3. You first need to create a Factor, which is essentially a user that will be accessing this service. The easiest way to do this is through a cURL command:
```
curl -X POST "https://verify.twilio.com/v2/Services/VAXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX/Entities/ff483d1ff591898a9942916050d2ca3f/Factors" \
--data-urlencode "FriendlyName=Taylor's Account Name" \
--data-urlencode "FactorType=totp" \
-u $TWILIO_ACCOUNT_SID:$TWILIO_AUTH_TOKEN
```
- `VAXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX` is the ID of the service you made, which you can find here:
<img width="1140" alt="Screen Shot 2023-03-12 at 2 10 07 PM" src="https://user-images.githubusercontent.com/64047863/224563981-781ea3af-b137-4568-a414-7cf6d3bd8990.png">

- `ff483d1ff591898a9942916050d2ca3f` is a unique ID that will be associated with your user. This can be a randomly generated from some magical website you find
- The FriendlyName is the name of your user which can be anything
- Don't change FactorType
- `$TWILIO_ACCOUNT_SID` and `$TWILIO_AUTH_TOKEN` are found under `Account > Auth tokens & API Keys`

4. After this hopefully works you will get a response like this:
```
{
  "sid": "YFXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX",
  "account_sid": "ACXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX",
  "service_sid": "VAXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX",
  "entity_sid": "YEXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX",
  "identity": "ff483d1ff591898a9942916050d2ca3f",
  "binding": {
    "secret": "GEZDGNBVGY3TQOJQGEZDGNBVGY3TQOJQ",
    "uri": "otpauth://totp/test-issuer:John%E2%80%99s%20Account%20Name?secret=GEZDGNBVGY3TQOJQGEZDGNBVGY3TQOJQ&issuer=test-issuer&algorithm=SHA1&digits=6&period=30"
  },
  "date_created": "2015-07-30T20:00:00Z",
  "date_updated": "2015-07-30T20:00:00Z",
  "friendly_name": "Taylor's Account Name",
  "status": "unverified",
  "factor_type": "totp",
  "config": {
    "alg": "sha1",
    "skew": 1,
    "code_length": 6,
    "time_step": 30
  },
  "metadata": null,
  "url": "https://verify.twilio.com/v2/Services/VAXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX/Entities/ff483d1ff591898a9942916050d2ca3f/Factors/YFXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX"
}
```
5. Now that the Factor is successfully created, you have to verify that they're able to generate and send the correct code. First, download and open the Authy Desktop App and hit the plus sign in the upper right corner of the app. In the box, enter the `binding.secret` token from the Factor creation, and name the application accordingly. Now, you should be able to see a code being generated.
6. Use the following cURL request to verify the user:
```
curl -X POST "https://verify.twilio.com/v2/Services/VAXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX/Entities/ff483d1ff591898a9942916050d2ca3f/Factors/YFXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX" \
--data-urlencode "AuthPayload=293412" \
-u $TWILIO_ACCOUNT_SID:$TWILIO_AUTH_TOKEN
```
- All fields are the same as above, with the exception of `YFXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX`. This can be found in the create Factor response, under `sid`
- The `AuthPayload` will be the code generated by the Authy app which you'll need to copy over from the desktop app
7. If you get a response that includes `status: approved`, you're done the setup! (once again I apologize this is so long, this only needs to be done once and moving forward all users will be under one general Blueprint account
8. Finally update your `.env` file with the following:
```
# Twilio
TWILIO_ENABLED=True
TWILIO_ACCOUNT_SID=Account SID you used before
TWILIO_AUTH_TOKEN=Auth token you used before
TWILIO_SERVICE_SID=Service SID you used before
TWILIO_ENTITY_ID=Entity ID of the user you created and used before
TWILIO_FACTOR_SID=Factor SID of the factor you made and used before
```
9. Test the 2FA process locally. Ensure that support users are brought through this page, while all other users are not.

If anything here doesn't make sense, you can [refer to the docs](https://www.twilio.com/docs/verify/quickstarts/totp)

## What should reviewers focus on?
* 


## Checklist
